src/smc-webapp/jupyter/widgets/manager.ts

```diff
@@ -115,16 +116,48 @@ export class WidgetManager extends base.ManagerBase<HTMLElement> {
     // Now store the model in the
   }
 
+  /*
+  Documention for IPY_MODEL_ reference spec:
+
+      https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/schema/jupyterwidgetmodels.v7-4.md#model-state
+
+  TODO: there's no way I can see how to tell which values will be references,
+  so we just search from everything for now, at least as a string (layout, style) or
+  array of strings (children, axes, buttons).  I wish I knew that those 5 were the
+  only keys to consider...  I'm also worried because what if some random value
+  just happens to look like a reference?
```

Comment:
We tried to mitigate this somewhat by prefixing widget link ids with "IPY_MODEL_", so you might also check for that prefix before.

The widget itself is in charge of understanding which bits of its state are widget links and dereferencing widget links using its custom deserializer, which understands how to convert JSON data to the object instance attributes. For example, for a box widget: https://github.com/jupyter-widgets/ipywidgets/blob/092c085aaf3b1165a5c8f24a9ce663ae523e7969/packages/controls/src/widget_box.ts#L84

Reply:
(source for that deserializer is in https://github.com/jupyter-widgets/ipywidgets/blob/092c085aaf3b1165a5c8f24a9ce663ae523e7969/packages/base/src/widget.ts#L35

Reply:
Hmm.  I was following the examples so used set(state) to set the state of the model, which is why I ended up having to basically re-implementing unpack_models (since I didn't know about it).  But instead I should use a deserialize method?  I'll see if I can get that to work.  If not, at least I will rename my function something involving "unpack_model" to clarify the connection.

My intention is NOT to use comm's at all for sync between browser clients and the kernel.  Instead that sync will use cocalc's realtime sync.  The comms will then only be used between cocalc's node.js backend and the jupyter kernel (all in the project).   This at least doesn't try to fight with anything in the comm implementation that assume that there is precisely one browser client.

Reply:
Great. I want to move to a centralized state paradigm, so I'm watching very closely to what you are doing :).

The deserializer dictionary is a static attribute on the widget model class. You might be able to just access and use those directly, no matter what widget model you have, so you don't have to reimplement every deserializer every widget uses and figure out where to use it. In fact, you may be able to just directly use the model class's static _deserialize_state method, like we do: https://github.com/jupyter-widgets/ipywidgets/blob/092c085aaf3b1165a5c8f24a9ce663ae523e7969/packages/base/src/manager-base.ts#L355

That function is defined here: https://github.com/jupyter-widgets/ipywidgets/blob/1f93ad39802d3ee4a6b88bc13064f93f8f796374/packages/base/src/widget.ts#L522